### ISSUE TITLE
UI size and resize miracle ;-)

### ISSUE THREAD

**Issue by sjaehn** [lv2/pugl#42]
After our yesterdays discussion I did some experiments. Your idea was NOT to call ui_resize() from LV2UI and just let the host find the right size. Makes sense. But I observed that the plugin UI is then showed with its min size. In the most cases. Without any need (full HD screen). I don't know if this is a feature or a bug. And if it is somehow related to pugl (I've some doubts). Or more likely to the hosts and/or their libraries (suil?). But then I would keep calling ui_resize() following puglRealize().

I used and manipulated B.Oops and pugl in jalv.gtk3, jalv.qt5, Carla, ardour 6.x, and qtractor. The relevant plugin code contains (https://github.com/sjaehn/BOops/blob/master/src/BWidgets/Window.cpp#L47):
```
	puglSetDefaultSize (view_, getWidth (), getHeight ());
	puglSetMinSize (view_, getWidth () / 4.0, getHeight () / 4.0);
	//...
	puglSetViewHint(view_, PUGL_RESIZABLE, PUGL_TRUE);
	puglSetViewHint(view_, PUGL_IGNORE_KEY_REPEAT, PUGL_TRUE);
	//...
	puglSetBackend(view_, puglCairoBackend());
	//...
	puglRealize (view_);
	puglShow (view_);
```
with getWidth() = 1240, and getHeight() = 608.

As I told you yesterday, all plugin UIs are shown in a largely reduced size in all five conditions. In the min size in jalv, Carla, and Ardour. A bit larger (640x480) in qtractor. But puglRealize calls XCreateWindow() with 1240x608. XGetWindowAttributes() confirmed the window size of 1240x608.

And there are NO PUGL_CONFIGURE events emitted in jalv.qt5. And not in qtractor. As @ventosus mentioned, the only way to inform the plugin UI about a resize in qtractor is via the ui:resize extension. I can confirm this. This works now :-).

I added a stderr output in the event function if a PUGL_CONFIGURE event is emitted ("configure ", window.x, window.y, window.width, window.height). Interesting output, just after puglRealize():

```
(jalv.gtk3:124693): Gtk-WARNING **: 13:24:15.244: SuilX11Wrapper 0x556f7720dba0 reported min size 152 and natural size 0 in get_preferred_height(); natural size must be >= min size

(jalv.gtk3:124693): Gtk-WARNING **: 13:24:15.245: SuilX11Wrapper 0x556f7720dba0 reported min size 310 and natural size 0 in get_preferred_width(); natural size must be >= min size
configure 0,000000 0,000000 1240,000000 608,000000

(jalv.gtk3:124693): Gtk-WARNING **: 13:24:15.340: SuilX11Wrapper 0x556f7720dba0 reported min size 152 and natural size 0 in get_preferred_height(); natural size must be >= min size

(jalv.gtk3:124693): Gtk-WARNING **: 13:24:15.341: SuilX11Wrapper 0x556f7720dba0 reported min size 310 and natural size 0 in get_preferred_width(); natural size must be >= min size
configure 0,000000 0,000000 310,000000 152,000000
```

The first PUGL_CONFIGURE was with the default size. The second one (100 ms later) with min size. 

My conclusion: The plugin UI is correctly realized by pugl (puglRealize) with the right default size. But the host tries to (brutally) downsize the plugin window to its min size. Finally, I would keep ui_resize() as I can successfully control the size in this way.

Just to inform you. I don't know if and where sth. should be changed / fixed.

**Comment by ventosus** [comment]
Does this help?

There seem to be missing some bit-wise ORs in constructing the size hint flags.

    sizeHints.flags      |= PMinSize;
    sizeHints.flags      |= PMaxSize;

https://github.com/lv2/pugl/blob/master/src/x11.c#L224
https://github.com/lv2/pugl/blob/master/src/x11.c#L229

**Comment by sjaehn** [comment]
BTW, in https://github.com/lv2/pugl/blob/master/src/x11.c#L1275 just saw this:
```
PuglStatus
puglSetMaxSize(PuglView* const view, const int width, const int height)
{
  view->minWidth  = width;
  view->minHeight = height;
  return updateSizeHints(view);
}
```
This can't be right (min vs max). But this didn't cause the problems.

**Comment by sjaehn** [comment]
> Does this help?
> 
> There seem to be missing some bit-wise ORs in constructing the size hint flags.
> 
> ```
> sizeHints.flags      |= PMinSize;
> sizeHints.flags      |= PMaxSize;
> ```
> 
> https://github.com/lv2/pugl/blob/master/src/x11.c#L224
> https://github.com/lv2/pugl/blob/master/src/x11.c#L229

@ventosus: Sorry, it doesn't help. But I've to take a closer look to this part of source code (later / next days). puglUpdateSizeHints is only called from puglRealize() (after XCreateWindow() ), puglSetMinSize(), puglSetMaxSize(), and puglSetAspectRatio() (BTW, aspect ratios are only set if (!view->hints[PUGL_RESIZABLE]) ?). But no call from puglSetViewHint.

**Comment by drobilla** [comment]
It is probably a bug in either suil, or hosts, or both. Possibly also pugl but it all seems to work for top-level applications.

Basically the situation here is that size stuff with embedding isn't implemented right, and nobody has looked into it or really fixed it yet (at least not entirely, there's been progress of course).  Plugin authors discover ui_resize() and hammer on that in inconsistent ways to make things do what they want, which causes a bunch of other problems.

In other words I wasn't suggesting that not using ui_resize() at all will fix everything in practice right now, but it's what we need to make work for this stuff to actually work properly.

Anyway, I'm traveling/moving right now so it'll be at least a few days before I can look into it, but I'll try.  Basically the values set in C (hopefully properly...) need to be traced through suil to the corresponding widget properties to see that everything is set sensibly.

**Comment by drobilla** [comment]
> BTW, in https://github.com/lv2/pugl/blob/master/src/x11.c#L1275 just saw this:

Oof 🤦

That bit fixed in e8ba623, thanks.

**Comment by falkTX** [comment]
To continue with this, I recently made Carla support LV2 UI resizing without involving extensions, tested with DPF develop branch example plugins.
Code in question https://github.com/falkTX/Carla/commit/75ea9f5b2d184151c696f1dc648dfb700959c876

Now to be done in suil.
Reaper also can't do this yet, but that one we cannot change/fix ourselves.

**Comment by falkTX** [comment]
For those that want to do some testing..
clone https://github.com/DISTRHO/DPF with develop branch.
edit https://github.com/DISTRHO/DPF/blob/87d4cac0fe404877fd34c5aeac51a833b065924a/examples/Info/DistrhoPluginInfo.h#L29 to say 0 instead of 1.

 "make" as usual to build the whole thing.
./bin/ inside the repo will contain the binaries (including lv2 bundles)

The "info" example plugin will show a resize handle to do plugin-side window resizing.

![Screenshot_20210531_125920](https://user-images.githubusercontent.com/1334853/120190087-07bca180-c210-11eb-87dd-f8abd886c529.png)

**Comment by drobilla** [comment]
This doesn't really seem like a Pugl issue, but the issues in the comments above have been fixed, and recent work like d8e02dad and 1cd37cad apply size hints more consistently, and 82405e8c adds useful API for plugins to change only the position or size.

So, I'll assume this is a non-issue and/or fixed by now.  Please open a more precise ticket if you still encounter any problems with window sizing.
